User notes

samarthsharma36 opened the pull request:
Task : 
CRUD on UserNotes

Steps To Test : 
1. create a user with name = "Super User" , username = 'admin' and password = 'admin'
2. create a category 'Uncategorised'.
Usually this category is default but just in case

Please Review and suggest changes :)  @puneet0191 @yvesh

wilsonge committed:
Use the helper function (#22)

samarthsharma36 committed:
User Notes CRUD

puneet0191 commented on src/_support/Page/Acceptance/Administrator/userNotesPage.php:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @package  Page\Acceptance\Administrator
+ *
+ * @since    3.7.3
```
since 4.0.0

puneet0191 commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$this->subject = "Test Subject";
+		$this->notes = 'This is a test note for user '.$this->username;
+	}
+	/**
```
a space here

puneet0191 reviewed (changes requested):
nice work with this one, I will merge once the changes requested are done

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->wait(1);
+		$I->click(['link' => $this->category]);
+		$I->switchToPreviousTab();
+		$I->wait(1);
```
We should try to avoid static timeouts like wait(x).
We have a 6s delay (6x wait(1)) in this method. 

We need to find workarounds for this. I.e.: wait for elements to become visible or wait for js to be executed. 
Otherwise the overall test execution time will be horrible.

samarthsharma36 replied:
I have tried them all but static wait is only working. I throws error :
 "message":"Cannot find context with specified id"

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @link      ArticleMenuCest
+ * @since     __DEPLOYED_VERSION
+ */
+class userNotesCest
```
class name in CamelCase

dneukirchen replied:
@puneet0191 dont we want to split list cests and form cests?

samarthsharma36 replied:
@puneet0191 suggested me to split only page files into list and form. He asked me not to split cest files

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @author    Samarth sharma <samarthsharma351@gmail.com>
+ * @copyright 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license   Joomla 2005-2018
+ * @link      ArticleMenuCest
```
!= ARTICLE

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @category  Menu_Article
+ * @package   Administratorcomponentscom_Menu_Article
+ * @author    Samarth sharma <samarthsharma351@gmail.com>
+ * @copyright 2005 - 2016 Open Source Matters, Inc. All rights reserved.
```
wrong year

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * Administrator UserNotes Tests
+ *
+ * @category  Menu_Article
+ * @package   Administratorcomponentscom_Menu_Article
```
Wrong package

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+/**
+ * Administrator UserNotes Tests
+ *
+ * @category  Menu_Article
```
probably wrong category

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @package     Joomla.Tests
+ * @subpackage  Acceptance.tests
+ *
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
```
wrong copyright

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ */
+
+namespace administrator\components\com_users;
+use Page\Acceptance\Administrator\userNotesPage as UserNotes;
```
nl btween namespace and use statements.

samarthsharma36 replied:
changed

dneukirchen commented on src/_support/Page/Acceptance/Administrator/userNotesPage.php:
```diff
@@ -0,0 +1,78 @@
+ *
+ * @since    3.7.3
+ */
+class userNotesPage
```
@puneet0191 dont we want to split list po´s and form po´s?

dneukirchen replied:
class names in CamelCase please

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @package   Administratorcomponentscom_Menu_Article
+ * @author    Samarth sharma <samarthsharma351@gmail.com>
+ * @copyright 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license   Joomla 2005-2018
```
wrong license

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+ * @copyright 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license   Joomla 2005-2018
+ * @link      ArticleMenuCest
+ * @since     __DEPLOYED_VERSION
```
since 4.0.0 or `__DEPLOYED_VERSION__`

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$this->name = "Super User";
+		$this->category = "Uncategorised";
+		$this->subject = "Test Subject";
+		$this->notes = 'This is a test note for user '.$this->username;
```
please use singular if its just one note.

dneukirchen replied:
Please add spaces around the dot `user '.$this->username`

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @param   \AcceptanceTester  $I The AcceptanceTester Object
+	 *
+	 * @since   3.7.3
```
wrong since

samarthsharma36 replied:
changed

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @return  void
+	 */
+	public function createUserNotes(\AcceptanceTester $I)
```
You are creating only one note. Please rename to `createUserNote` (singular).

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @param   \AcceptanceTester  $I The AcceptanceTester Object
+	 *
+	 * @since   3.7.3
```
wrong since

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->fillField(UserNotes::$editor,$this->notes);
+		// Save and close
+		$I->click(UserNotes::$dropDownToggle);
+		$I->clickToolbarButton('save & close');
```
Im missing the assertion part of the test

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->amOnPage(UserNotes::$url);
+		$I->searchForItem($this->subject);
+		$I->click(UserNotes::$option1);
+		$I->clickToolbarButton('unpublish');
```
Im missing the assertion part of the test

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @param   \AcceptanceTester  $I The AcceptanceTester Object
+	 *
+	 * @since   3.7.3
```
wrong since

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->amOnPage(UserNotes::$url);
+		$I->searchForItem($this->subject);
+		$I->click(UserNotes::$option1);
+		$I->clickToolbarButton('trash');
```
Im missing the assertion part of the test

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->amOnPage(UserNotes::$url);
+		$I->searchForItem($this->subject);
+		$I->click(UserNotes::$option1);
+		$I->clickToolbarButton('check-in');
```
Im missing the assertion part of the test

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @param   \AcceptanceTester  $I The AcceptanceTester Object
+	 *
+	 * @since   3.7.3
```
wrong since

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+		$I->amOnPage(UserNotes::$url);
+		$I->searchForItem($this->subject);
+		$I->click(UserNotes::$option1);
+		$I->clickToolbarButton('publish');
```
Im missing the assertion part of the test

dneukirchen commented on src/acceptance/administrator/components/com_users/userNotesCest.php:
```diff
@@ -0,0 +1,145 @@
+	 *
+	 * @param   \AcceptanceTester  $I The AcceptanceTester Object
+	 *
+	 * @since   3.7.3
```
wrong since

dneukirchen commented:
Thank you @Samgithub1997 for the contribution and sorry for the large number of review comments (most of them are code style related), but imo this PR is not mergeable i its current state. 

* Most of the tests are missing the assertion part of the test.
* We should avoid static timeouts like wait(x)
* We should split cests and po´s in list and form classes and use inheritance and composition to abstract common tasks.

samarthsharma36 committed:
form amd list pages

samarthsharma36 committed:
test assertions added

samarthsharma36 committed:
Update userNotesCest.php

samarthsharma36 committed:
Update userNotesCest.php

samarthsharma36 commented:
@dneukirchen @puneet0191 
Please review the last commit for User-notes. I have made the necessary changes,  
except the static wait() part because I have tried waitForElement , waitForElementNotVisible, waitForElementVisible and waitForJS. But none worked so I kept the static wait to min
$I->wait(.2) . So there will only be a delay of 1.2 secs in whole test.
CHANGES : 
1. Seperate listPage and formPage file
2. docblock comment
3. Assertion of tests

dneukirchen commented:
Hey @Samgithub1997 
Thanks for the updates. Not all of my issues are resolved: 

* Unfortunately this is not what i meant with "assertion". You just added a "I wantToTest" comment. In general tests consist of 3 parts: setup/arrangement, execution and assertion (others call those given, when, then) and your tests are missing the last part. You are only executing certain actions, but you do not check that they are successful. i.e.: https://github.com/Samgithub1997/test-system/blob/e3a2254a888ffb139a43501f708f267d53ff4f9b/src/acceptance/administrator/components/com_users/userNotesCest.php#L88 here you want to test, that it is possible to unbublish user notes, but you are not checking that the note was unpublished successfully. 
* Thanks for testing the static timeout issue. I still believe that there is a solution for it.
* We have still some code style and naming convention issues (i.e. Non CamelCase Class- and Filenames, wrong usage of namespaces, wrong DocBlocks, Joomla Codestyle in general). Please discuss that with your mentors and look at the common resources https://developer.joomla.org/coding-standards.html

samarthsharma36 committed:
spaces and camelCase

samarthsharma36 committed:
spaces, camelCase and static wait

samarthsharma36 committed:
Merge branch 'user-notes' of https://github.com/Samgithub1997/test-system into user-notes

samarthsharma36 committed:
indentation

Assertion is left

samarthsharma36 committed:
assertion

samarthsharma36 committed:
Merge branch 'user-notes' of https://github.com/Samgithub1997/test-system into user-notes

samarthsharma36 committed:
Assertions added

samarthsharma36 committed:
Update UserNotesListPage.php

samarthsharma36 committed:
Update UserNotesFormPage.php

samarthsharma36 commented:
@dneukirchen @puneet0191  
=> I have changed filenames
=> I am not quite sure if the assertions I have added is what you require but please let me know the changes
=> Added lines btwn two functions
https://github.com/joomla/test-system/pull/41/files#diff-3ebae07201aca9bcd0ba86c50217ea12

puneet0191 commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+namespace administrator\components\com_users;
+use Page\Acceptance\Administrator\UserNotesListPage as UserNotesList;
+use Page\Acceptance\Administrator\UserNotesFormPage as UserNotesForm;
+/**
```
a line here

puneet0191 commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+		$I->amOnPage(UserNotesList::$url);
+		$I->searchForItem($this->subject);
+		$I->click(UserNotesList::$option1);
+		$I->clickToolbarButton('unpublish');
```
this is missing assertion, now when you unpublish the user Note, you must verify if the status is actually changing or not, so basically either use $I->see or $I->dontSee on the status for the note.

samarthsharma36 replied:
Added Assertion

puneet0191 commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+		$this->name = "Super User";
+		$this->category = "Uncategorised";
+		$this->subject = "Test Subject";
+		$this->editorText = 'This is a test note for user '. $this->username;
```
have an empty space here, basically something like this. 
`$this->editorText = 'This is a test note for user ' . $this->username;`

puneet0191 replied:
@Samgithub1997  why is this left?

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+ * @link      User Notes
+ * @since     4.0.0
+ */
+class userNotesCest
```
Class names must be CamelCase

puneet0191 replied:
@Samgithub1997  why are you not looking here?

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,75 @@
+	 * @var    string
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public static $url = '/administrator/index.php?option=com_users&view=notes';
```
Is this the url of the User Notes Form(!) Page?

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace administrator\components\com_users;
```
Usually we do not namespace the cests, do we?

samarthsharma36 replied:
Yeah. You are right.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+
+namespace administrator\components\com_users;
+use Page\Acceptance\Administrator\UserNotesListPage as UserNotesList;
+use Page\Acceptance\Administrator\UserNotesFormPage as UserNotesForm;
```
Why do we use aliases here?

samarthsharma36 replied:
Aliases are very helpful as when I was asked to change the name of file from userNotesListPage to UserNotesListPage. I only had to make the change once

dneukirchen replied:
Thats not a good reason to use aliases. 
Aliases reduce readability, so we should only use them in case of class name conflicts.

samarthsharma36 replied:
Ok. I have removed aliases in my latest commit

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+		$I->clickToolbarButton('publish');
+		// Assertion
+		$I->searchForItem($this->subject);
+		$I->see($this->subject);
```
This does not check that the note is published.

samarthsharma36 replied:
I will add the status on my next commit 
![selection_001](https://user-images.githubusercontent.com/19748816/42033337-35080f18-7afa-11e8-8aa2-d21a1f0da5b7.png)

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+		$I->clickToolbarButton('check-in');
+		// Assertion
+		$I->searchForItem($this->subject);
+		$I->see($this->subject);
```
This does not check that the note is checked in.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,162 @@
+		$I->clickToolbarButton('trash');
+		// Assertion
+		$I->searchForItem($this->subject);
+		$I->dontSee($this->subject);
```
This does not check that the note is trashed.

samarthsharma36 replied:
It does, it removes the item from backend admin

dneukirchen reviewed (changes requested):
thx @Samgithub1997, you are getting there.

samarthsharma36 committed:
Indentations, docblock and Aassertions

samarthsharma36 committed:
Merge branch 'user-notes' of https://github.com/Samgithub1997/test-system into user-notes

samarthsharma36 committed:
Update UserNotesListPage.php

samarthsharma36 committed:
indent and docblock comment

samarthsharma36 commented:
@puneet0191 @dneukirchen  please review. I tried to cover all the issues, :)
Left : Assertion of user-note checked in

dneukirchen commented:
There are still unresolved review comments.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,71 @@
+ * @package   Page\Acceptance\Administrator
+ * @author    Samarth sharma <samarthsharma351@gmail.com>
+ * @copyright 2005 - 2018 Open Source Matters, Inc. All rights reserved.
+ * @license   Joomla 2005-2018
```
Still wrong license

samarthsharma36 replied:
done

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @category  Users
+ * @package   Page\Acceptance\Administrator
+ * @author    Samarth sharma <samarthsharma351@gmail.com>
```
Please remove author.  Only permitted in non-Joomla source files

samarthsharma36 replied:
done

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+ * @since      4.0.0
+ */
+use Page\Acceptance\Administrator\UserNotesListPage as UserNotesList;
+use Page\Acceptance\Administrator\UserNotesFormPage as UserNotesForm;
```
Please remove aliases as the class names are more explicit.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+		$I->wantToTest('that user notes are unpublished.');
+		$I->doAdministratorLogin();
+		$I->amOnPage(UserNotesList::$url);
+		$I->searchForItem($this->subject);
```
This test and the following tests depend on createUserNote test, which in general is not good. Please discuss this with your mentors. I would prefer to write independent tests. If you dont want to go that way, then please at least mark those tests with  `@depends xxx` annotation.

puneet0191 replied:
yes, that is correct, @Samgithub1997  use @depends annotation for the test depending on the data created in the previous tests, 
https://codeception.com/docs/07-AdvancedUsage#Dependencies

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+		$I->clickToolbarButton('unpublish');
+		$I->searchForItem($this->subject);
+		// Status changes
+		$I->seeElement(UserNotesList::$unpublishStatus);
```
This is not bulletproof. 

1) There might be other unpublished notes
2) Filter settings might hide the note

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+		// Assertion
+		$I->searchForItem($this->subject);
+		// Status changes
+		$I->seeElement(UserNotesList::$publishStatus);
```
This is not bulletproof. 

1) There might be other published notes
2) Filter settings might hide the note

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+	 */
+	public function publishUserNote(\AcceptanceTester $I)
+	{
+		$I->wantToTest('that user note is published');
```
The note is only unpublished when the previous test was successful. So this tests kind of depends on 2 previous tests.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+		// Assertion
+		$I->searchForItem($this->subject);
+		// Status changes
+		$I->seeElement(UserNotesList::$publishStatus);
```
This does not check that the item is checked in.

puneet0191 reviewed (changes requested):
Please review Comment correctly.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+		// Assertion
+		$I->searchForItem($this->subject);
+		// Item is not visible after deletion
+		$I->dontSee($this->subject);
```
This does not check that the item is in trash.

samarthsharma36 replied:
The item is not available on backend after trash. so how does it not check that?

dneukirchen replied:
It is. Its only soft deletion. Item is in trash !== deleted. 
Please check the trashed filter.

samarthsharma36 replied:
Got it. Thanks

samarthsharma36 committed:
Assertions + liscense

samarthsharma36 commented:
I still think use of Aliases is a better idea
I added Assertions with filter. Please review when you get free :)
@puneet0191  @dneukirchen

samarthsharma36 committed:
alias

samarthsharma36 commented:
Removed Alias

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -17,21 +14,20 @@
 		$this->name = "Super User";
 		$this->category = "Uncategorised";
-		$this->subject = "Test Subject";
+		$this->subject = "Test Subject 999";
```
Why 999?

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -68,103 +64,55 @@ public function createUserNote(\AcceptanceTester $I)
 	 * @return  void
 	 */
-	public function unpublishUserNote(\AcceptanceTester $I)
+	public function toolbarOperationsUserNote(\AcceptanceTester $I)
```
Why did you change the name and moved everything into one test?
We should keep the tests als small as possible (in my own projects i have the rule of thumb that a test should not be longer than 10-15).

When this test fails, you have to look into the logs to know which part of the test failed.

samarthsharma36 replied:
Because all tests were dependent.

dneukirchen replied:
But moving everything that depends on something else into one massive test function is not the solution. Imo it is even worse than having tests that depend on each other.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -44,7 +40,7 @@ public function __construct()
-	public function createUserNote(\AcceptanceTester $I)
+	public function createUserNotes(\AcceptanceTester $I)
 	{
 		$I->wantToTest('creating user notes ');
```
whitespace after "notes"

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -68,103 +64,55 @@ public function createUserNote(\AcceptanceTester $I)
 		$I->scrollTo(UserNotesForm::$toggleEditor);
 		$I->click(UserNotesForm::$toggleEditor);
-		$I->fillField(UserNotesForm::$editor, $this->editorText);
+		$I->fillField(UserNotesForm::$editor,$this->editorText);
```
Please reverrt this. Whitespace after comma was good.

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -68,103 +64,55 @@ public function createUserNote(\AcceptanceTester $I)
 
 	/**
-	 * Unpublish User Notes
+	 * Toolbar Operations On User Notes
```
see below.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,49 @@
+     * @var    array
+     * @since  4.0.0
+     */
+    public static $unpublishStatus = ['xpath' => '//span[@class="icon-unpublish"]'];
```
Please ALWAYS! remove unused code.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,49 @@
+     * @var    array
+     * @since  4.0.0
+     */
+    public static $publishStatus = ['xpath' => '//span[@class="icon-publish"]'];
```
Please ALWAYS! remove unused code.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,49 @@
+     * @var    array
+     * @since  4.0.0
+     */
+    public static $option1 = ['id' => 'cb0'];
```
Can we find a better name for this locator?

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,118 @@
+ * Administrator UserNotes Tests
+ *
+ * @category  Users
+ * @package   Administratorcomponentscom_Users
```
Wrong package (package is optional)

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,118 @@
+ * @package   Administratorcomponentscom_Users
+ * @copyright 2005 - 2018 Open Source Matters, Inc. All rights reserved.
+ * @license   GNU General Public License version 2 or later; see LICENSE.txt
+ * @link      User Notes
```
Is this a link?

samarthsharma36 replied:
Removed.

dneukirchen commented:
Dont know, why gh marks my first 5 comments as outdated. They are not. Please check them

samarthsharma36 committed:
link rm

samarthsharma36 committed:
Update UserNotesCest.php

samarthsharma36 commented:
whats GH @dneukirchen

dneukirchen commented:
github

samarthsharma36 committed:
removed unused code

samarthsharma36 committed:
Merge branch 'user-notes' of https://github.com/Samgithub1997/test-system into user-notes

samarthsharma36 commented:
I checked your previous comments. Which one isn't covered? @dneukirchen

dneukirchen commented:
The first 6 of my last review

samarthsharma36 committed:
spaces

samarthsharma36 committed:
spaces

samarthsharma36 committed:
indentation

samarthsharma36 committed:
depends

samarthsharma36 committed:
Update UserNotesCest.php

samarthsharma36 commented:
@puneet0191 @dneukirchen I have made seperate functions now

puneet0191 commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,69 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $selectUserButton = ['xpath' => '//*[@title="Select User"]'];
```
avoid using * in xpaths, also could you please change this selector to something like this: 
`public static $selectUserButton = "//*[@title='Select User']";`
but please ensure you replace * used in xpath with the actuall html element, could be a div, p, button etc.

puneet0191 commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,69 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $editor = ['id' => 'jform_body'];
```
change the id based selector to something like this: 
`public static $editor = "#jform_body";`

puneet0191 commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @var    array
+     * @since  4.0.0
+     */
```
I think you are missing something here, what is this block for?

samarthsharma36 replied:
I don't know if I have missed any but I followed exactly according to developer.joomla.org 
![selection_006](https://user-images.githubusercontent.com/19748816/42150229-bbfbee64-7df6-11e8-99f1-d619d6b847f3.png)

puneet0191 reviewed (changes requested):
@Samgithub1997  comments about the Page object, fix them please, good progress on the improvements here, keep it up.

samarthsharma36 committed:
ids and xpath

samarthsharma36 commented:
@puneet0191  , I made changes in xpath and ids.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,69 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $subject = '#jform_subject';
```
Not an array

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,69 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $selectCategory = '#jform_catid_select';
```
Not an array

puneet0191 replied:
@dneukirchen you mean `@var` should be changed to string, because I asked him to change these to String, to ensure compatibility with codeception 2.4.x

dneukirchen replied:
yes, only the docblock is wrong now.

samarthsharma36 replied:
added string now

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesFormPage.php:
```diff
@@ -0,0 +1,69 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $editor = '#jform_body';
```
Not an array

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,32 @@
+ * @package     Joomla.Test
+ * @subpackage  AcceptanceTester.Page
+ *
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
```
wrong year

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,32 @@
+class UserNotesListPage
+{
+	// Include url of current page
+	public static $url = '/administrator/index.php?option=com_users&view=notes';
```
Missing docblock

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,32 @@
+	/**
+	 * Option one in list
+	 *
+	 * @var    array
```
Not an array.

dneukirchen commented on src/_support/Page/Acceptance/Administrator/UserNotesListPage.php:
```diff
@@ -0,0 +1,32 @@
+	 * @var    array
+	 * @since  4.0.0
+	 */
+	public static $optionOne = '#cb0';
```
not exactkly what i had in mind, when i asked you to rename to something more explicit. 
What do you think about `checkboxOfFirstListItem`?

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+	 *
+	 * @return  void
+	 */
+	public function unpublishUserNote (\AcceptanceTester $I)
```
Missing the @depends annotation that we talked about

puneet0191 replied:
@depends is mentioned, see here: https://github.com/joomla/test-system/pull/41/files/43c09b80aafe9448b77398f4170899c34dc3d493#diff-aa63769d16d496c60b7d7c19a78b9caeR77

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+	 *
+	 * @return  void
+	 */
+	public function publishUserNote (\AcceptanceTester $I)
```
Missing the @depends annotation that we talked about

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+	 *
+	 * @return  void
+	 */
+	public function archiveUserNote (\AcceptanceTester $I)
```
Missing the @depends annotation that we talked about

dneukirchen commented on src/acceptance/administrator/components/com_users/UserNotesCest.php:
```diff
@@ -0,0 +1,170 @@
+	 *
+	 * @return  void
+	 */
+	public function trashUserNote (\AcceptanceTester $I)
```
Missing the @depends annotation that we talked about

dneukirchen reviewed (changes requested):
Good job. Only some minor issues and one recommendation:

When you find yourself repeating the same code in multiple tests it is always an indicator that you can abstract some piece of code, improve readability and reduce the iverall number of lines.

samarthsharma36 committed:
docblock string

samarthsharma36 committed:
Update UserNotesCest.php

samarthsharma36 committed:
Update UserNotesCest.php

samarthsharma36 commented:
@dneukirchen I changed the name of variable and also changed the var type from array to string

samarthsharma36 commented:
Sorry, that was by accident

dneukirchen reviewed (approved):
GJ @Samgithub1997

samarthsharma36 committed:
Update UserNotesFormPage.php

puneet0191 commented:
Thank you @Samgithub1997
